neat_ml/tests/test_plotting.py

```diff
@@ -0,0 +1,379 @@
+                 y_col="PEO 8 kg/mol (wt%)",
+                 phase_col="Phase",
+                 xrange=[0, 20], yrange=[0, 20]),
+        ),
+        (
+            pmf.binodal_model,
+            "mathematical_model.png",
+            dict(x_col="Sodium Citrate (wt%)",
+                 y_col="PEO 8 kg/mol (wt%)",
+                 phase_col="Phase",
+                 xrange=[0, 20], yrange=[0, 20]),
+        ),
+    ],
+)
+def test_visual_regression_on_helpers(
+    tmp_path: Path,
+    baseline_dir: Path,
+    synthetic_df: pd.DataFrame,
+    writer: Callable,
+    fname: str,
+    extra_kwargs: dict,
+):
+    
+    csv = tmp_path / "input.csv"
+    synthetic_df.to_csv(csv, index=False)
+
+    if writer.__name__ == "binodal_model":
+        json_file_path = tmp_path / "test_params.json"
+        test_params = {
+          "MODEL_A": 0.955,
+          "MODEL_B": -5.73,
+          "MODEL_C": 581
+        }
+        with open(json_file_path, 'w') as f:
+            json.dump(test_params, f)
+        extra_kwargs["json_path"] = str(json_file_path)
+
+    out_png = tmp_path / "out" / fname
+    writer(file_path=csv, output_path=out_png, **extra_kwargs)
+ 
+    result = compare_images(
+        str(baseline_dir / fname), 
+        str(out_png), 
+        tol=2e-2)
+    assert result is None
+
+def test_plot_two_scatter_visual_regression(
+    tmp_path: Path,
+    baseline_dir: Path,
+):
+    csv1 = tmp_path / "scatter_A.csv"
+    pd.DataFrame({"X": [1, 2, 3], "Y": [3, 2, 1]}).to_csv(csv1, index=False)
+
+    csv2 = tmp_path / "scatter_B.csv"
+    pd.DataFrame({"X": [1, 2, 3], "Y": [1, 2, 3]}).to_csv(csv2, index=False)
+
+    out_png = tmp_path / "out" / "plot_two_scatter.png"
+    pmf.plot_two_scatter(
+        csv1_path=csv1,
+        csv2_path=csv2,
+        output_path=out_png,
+        xlim=[0, 4],
+        ylim=[0, 4],
+    )
+    result = compare_images(
+        str(baseline_dir/ "plot_two_scatter.png"), 
+        str(out_png), 
+        tol=2e-2)
+    assert result is None
+
+@pytest.mark.parametrize("json_file, out_dict, err_msg, err_type",
+    [
+        (
+            "bad_params.json",
+            {"MODEL_A": 0.1, "MODEL_B": 0.0},
+            "is missing required keys: {\'MODEL_C\'}",
+            KeyError,
+        ),
+        (
+            "bad_value_type.json",
+            {"MODEL_A": 10.5, "MODEL_B": "not-a-number", "MODEL_C": 20.0},
+            "Parameter 'MODEL_B' must be a number, but got str.",
+            TypeError,
+        ),
+    ]
+)
+def test_load_json_errors(tmp_path: Path,
+json_file, out_dict, err_msg, err_type):
+    """
+    The JSON is valid but omits MODEL_C -> KeyError expected.
+    
+    AND	
+
+    Tests that a TypeError is raised if a parameter value is not a number.
+    
+    The JSON file contains all required keys, but the value for 'MODEL_B'
+    is a string, which should trigger a TypeError.
+    """
+    out_file = tmp_path / json_file
+    json.dump(out_dict, out_file.open("w"))
```

Comment:
it is good practice in Python to open a file using a context manager so the resource is closed in a timely manner